src/containers/containers.c

```diff
@@ -289,6 +293,340 @@ extern inline container_t *container_andnot(
+        case BITSET_CONTAINER_TYPE: {
+            const bitset_container_t *bc = const_CAST_bitset(c);
+            uint32_t wordindex = it->in_container_index / 64;
+            uint64_t word = bc->words[wordindex] &
+                            (UINT64_MAX << (it->in_container_index % 64));
+            do {
+                // Read set bits.
+                while (word != 0 && consumed < count) {
+                    *buf = high16 |
+                           (wordindex * 64 + roaring_trailing_zeroes(word));
+                    word = word & (word - 1);
+                    buf++;
+                    consumed++;
+                }
+                // Skip unset bits.
+                while (word == 0 &&
+                       wordindex + 1 < BITSET_CONTAINER_SIZE_IN_WORDS) {
+                    wordindex++;
+                    word = bc->words[wordindex];
+                }
+            } while (word != 0 && consumed < count);
+
+            it->has_value = word != 0;
+            if (it->has_value) {
+                it->in_container_index =
+                    wordindex * 64 + roaring_trailing_zeroes(word);
+                it->value = it->in_container_index;
+            }
+            return consumed;
+        }
+        case ARRAY_CONTAINER_TYPE: {
+            const array_container_t *ac = const_CAST_array(c);
+            uint32_t num_values =
+                minimum_uint32(ac->cardinality - it->in_container_index, count);
+            for (uint32_t i = 0; i < num_values; i++) {
+                buf[i] = high16 | ac->array[it->in_container_index + i];
+            }
+            buf += num_values;
+            consumed += num_values;
+            it->in_container_index += num_values;
+            it->has_value = it->in_container_index < ac->cardinality;
+            if (it->has_value) {
+                it->value = ac->array[it->in_container_index];
+            }
+            return consumed;
+        }
+        case RUN_CONTAINER_TYPE: {
+            const run_container_t *rc = const_CAST_run(c);
+            do {
+                uint32_t largest_run_value = rc->runs[it->run_index].value +
+                                             rc->runs[it->run_index].length;
+                uint32_t num_values = minimum_uint32(
+                    largest_run_value - it->value + 1, count - consumed);
+                for (uint32_t i = 0; i < num_values; i++) {
+                    buf[i] = high16 | (it->value + i);
+                }
+                it->value += num_values;  // TODO: can this overflow?
+                buf += num_values;
+                consumed += num_values;
+
+                if (it->value > largest_run_value || it->value == 0) {
+                    it->run_index++;
+                    if (it->run_index < rc->n_runs) {
+                        it->value = rc->runs[it->run_index].value;
+                    } else {
+                        it->has_value = false;
+                    }
+                }
```

Comment:
Is `count == 0` valid? If so, `it->value` could start at 0, `num_values = min(..., 0) = 0`, so `it->value = 0 + 0`, this will then consume all the ranges, and leave the iterator with `has_value = false` Should probably either document that count must be non-zero, or add a test to validate that calling with count == 0 leaves the iterator untouched.

Reply:
Added a check and test for `count == 0`.